libptable/ptable.c

```diff
@@ -0,0 +1,202 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Partition table
+ *
+ * Copyright 2020, 2023 Phoenix Systems
+ * Author: Hubert Buczynski, Lukasz Kosinski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <ctype.h>
+#include <endian.h>
+#include <string.h>
+
+/* Per project board configuration may overwrite PTABLE_CHECKSUM definition */
+#include <board_config.h>
+
+#include "ptable.h"
+
+
+/* Enable checksum for partition table */
+#ifndef PTABLE_CHECKSUM
+#define PTABLE_CHECKSUM 0
+#endif
+
+
+static uint32_t ptable_crc32(uint32_t crc, const void *data, size_t len)
+{
+	int i;
+
+	while (len--) {
+		crc ^= *(char *)data++;
+		for (i = 0; i < 8; i++) {
+			crc = (crc >> 1) ^ ((crc & 1) ? 0xedb88320 : 0);
+		}
+	}
+
+	return crc;
+}
+
+
+static int ptable_partVerify(const ptable_t *ptable, const ptable_part_t *part, uint32_t memsz, uint32_t blksz)
+{
+	const ptable_part_t *p;
+	int i;
+
+#if PTABLE_CHECKSUM
+	/* Verify partition checksum */
+	if (part->crc != ptable_crc32(0, part, sizeof(ptable_part_t) - sizeof(uint32_t))) {
```

Comment:
Guessing by the `0xedb88320` poly, reversed ITU-T V.42 CRC32 is used here, so initial crc value should be `(uint32_t)-1` a.k.a. `0xffffffff`, and not `0` - *this helps to give an output value other than `0` for an input string of  `0` octets*. Unless result of `ptable_crc32(..)` is not the partial result feeded into next partial update calculated with `ptable_crc32(prev_partial_crc, ..)`, but is the final result it needs to be exclusive-OR'ed the calculated CRC value with 0xffffffff (this is identical to doing a binary not on the CRC value), eg:
```suggestion
	uint32_t crc = ~ptable_crc32((uint32_t)-1, part, sizeof(ptable_part_t) - sizeof(uint32_t));
	if (part->crc != crc) {
```
To avoid multiplying the wrong usage a wrapper like below may be used, or the function `ptable_crc32(..)` modified directly (if partial check sums are not needed).
```C
static uint32_t ptable_crc32calc(const void *data, size_t len)
{
    return ~ptable_crc32((uint32_t)-1, data, len);
}
```
I refer to https://github.com/Michaelangel007/crc32 or https://en.wikipedia.org/wiki/Cyclic_redundancy_check  where it is pretty well explained.

Reply:
The original code was based on usage in jffs2 (maybe it's incorrect there?).
I took a look at the links you provided and calculated checksum for string `"123456789"` (test string used in the first link).
The original code returned incorrect `0x2dfd2d88`. After applying proposed changes we get correct `0xcbf43926`.